raft topology: add garbage collection for internal CDC generations table

patjed41 opened the pull request:
We add garbage collection for the `CDC_GENERATIONS_V3` table to prevent it from endlessly growing. This mechanism is especially needed because we send the entire contents of `CDC_GENERATIONS_V3` as a part of the group 0 snapshot.

The solution is to keep a clean-up candidate, which is one of the already published CDC generations. The CDC generation publisher introduced in #15281 continually uses this candidate to remove all generations with timestamps not exceeding the candidate's and sets a new candidate when needed.

We also add `test_cdc_generation_clearing.py` that verifies this new mechanism.

Fixes #15323

patjed41 commented:
v2: added a `!gen_rows->empty()` check to `get_cdc_generations_cleanup_candidate`

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1246,9 +1282,11 @@ class topology_coordinator {
                 co_await publish_oldest_cdc_generation(std::move(guard));
 
+                guard = co_await start_operation();
+                co_await clean_obsolete_cdc_generations(std::move(guard));
```
Can we do both under one guard (one write)?

patjed41 replied:
We can but it will require some refactoring.

One way to do this is making `publish_oldest_cdc_generation` take guard by const reference and return mutation.

Another way is making `publish_oldest_cdc_generation` and `clean_obsolete_cdc_generations` take guard and vector of mutations by reference. Then, we could update the topology state directly in the publisher's loop.

kbr-scylla replied:
Ah, I didn't see your comment.

Still, I like two guards more TBH.

gleb-cloudius replied:
Two guards are two barriers. Two chances to fail. The only advantage is two history rows, but we can be smart and build one entry that described what was done exactly.

kbr-scylla replied:
Ok. We could build an update string, like we build the update vector.

patjed41 replied:
Building the `reason` string may be a detail but I'll ask since I'm not sure how to approach it. I think we would like this building process to meet 2 requirements:
- the string should always look naturally (without something like ", " at the end),
- it shouldn't depend on the order in which `clean_obsolete_cdc_generations`,  `publish_oldest_cdc_generation` and possible future functions are called (unless they impose some order, but right now they don't).

For now, I implemented the solution that only meets the first requirement because meeting both of them requires more logic and I'm not sure if it's worth it.

kbr-scylla replied:
I think it's fine.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_clearing.py:
```diff
@@ -0,0 +1,98 @@
+async def test_dependency_on_timestamps(manager: ManagerClient):
+    """Test that CDC generations are not removed from CDC_GENERATIONS_V3 when the difference between their timestamp
+       and the topology coordinator's clock is too small. Then, test that the CDC generation publisher removes
+       the clean-up candidate (together with older generations) if its timestamp is old enough."""
```
Nice.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_clearing.py:
```diff
@@ -0,0 +1,98 @@
+
+        return (new_mark, new_gen_ids)
+
+    # The first generation should not be removed.
```
Please extend the comment, explaining the reason

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_clearing.py:
```diff
@@ -0,0 +1,98 @@
+    logger.info("Bootstrapping second node")
+    servers += [await manager.server_add()]
+
+    # Both generations should not be removed.
```
As above

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_clearing.py:
```diff
@@ -0,0 +1,98 @@
+    logger.info("Bootstrapping third node")
+    servers += [await manager.server_add()]
+
+    # The first generation should be removed.
```
This one is much more obvious, but could add something like " thanks to the above injection"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3680/

patjed41 commented:
v3:
- changed `clean_obsolete_cdc_generations` and `publish_oldest_cdc_generation` to be performed under one group 0 guard
- refactored `publish_oldest_cdc_generation` and `cdc_generation_publisher_fiber` to achieve the change above (in a new commit, the third one)
- changed `DISTINCT` to `LIMIT 1` in `req` in `system_keyspace::get_cdc_generations_cleanup_candidate`
- made comments in `test_cdc_generation_clearing.py` more expressive

kbr-scylla commented:
Looks like CI hanged:
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3697/
> Started 7 hr 17 min ago

last log lines:
> 11:58:48  [26413/26415] topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_tablets.2 190.06s
> 11:59:27  [26414/26415] topology_experimental_raft  debug  [ PASS ] topology_experimental_raft.test_tablets.3 191.22s

The new test has passed:
> [2023-09-15T09:49:58.173Z] [26206/26415] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_cdc_generation_clearing.1 6.90s
[2023-09-15T09:49:58.173Z] [26207/26415] topology  debug  [ PASS ] topology.test_topology_rejoin.1 34.42s
[2023-09-15T09:49:58.173Z] [26208/26415] topology  debug  [ PASS ] topology.test_snapshot.3 34.82s
[2023-09-15T09:49:58.173Z] [26209/26415] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_cdc_generation_clearing.2 5.91s
[2023-09-15T09:50:00.215Z] [26210/26415] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_blocked_bootstrap.2 15.21s
[2023-09-15T09:50:01.518Z] [26211/26415] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_cdc_generation_data.1 3.02s
[2023-09-15T09:50:01.518Z] [26212/26415] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_blocked_bootstrap.3 13.23s
[2023-09-15T09:50:04.369Z] [26213/26415] topology   dev   [ PASS ] topology.test_snapshot.3 208.76s
[2023-09-15T09:50:05.000Z] [26214/26415] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_cdc_generation_clearing.3 7.07s

so either some other test has hanged, or the framework has a bug that causes it to hang.
It's a known issue: https://github.com/scylladb/scylladb/issues/14465

I'm restarting CI

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3697/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3705/
